File: Tutorials/Codes/Modern_Control_Course/Chapter16/Lesson4/Chapter16_Lesson4.m
% Chapter16_Lesson4.m
% Example systems in controllable canonical form (CCF).
%
% Requires only base MATLAB for the from-scratch computations.
% The optional ss/tf/ctrb section requires Control System Toolbox.

clear; clc;

examples = {
    'Example 1: G1(s) = 2/(s^2 + 3s + 2)', [1 3 2], [2];
    'Example 2: G2(s) = (s + 2)/(s^3 + 6s^2 + 11s + 6)', [1 6 11 6], [1 2];
    'Example 3: G3(s) = (0.5s^2 + 1.5s + 1)/(s^3 + 4s^2 + 5s + 2)', [1 4 5 2], [0.5 1.5 1.0]
};

for k = 1:size(examples, 1)
    name = examples{k, 1};
    den = examples{k, 2};
    num = examples{k, 3};

    [A, B, C, D] = ccf_from_tf(den, num);
    Wc = controllability_matrix(A, B);

    fprintf('\n%s\n', repmat('=', 1, 72));
    disp(name);
    disp('A ='); disp(A);
    disp('B ='); disp(B);
    disp('C ='); disp(C);
    disp('D ='); disp(D);
    disp('Wc ='); disp(Wc);
    fprintf('rank(Wc) = %d\n', rank(Wc));
    fprintf('det(Wc)  = %.6g\n', det(Wc));

    if exist('ss', 'file') == 2 && exist('tf', 'file') == 2
        sys = ss(A, B, C, D);
        fprintf('Transfer function from ss(A,B,C,D):\n');
        tf(sys)
    end
end

% Optional Simulink generation: creates a simple continuous-time state-space
% block model for Example 1 if Simulink is available.
if exist('new_system', 'file') == 2
    [A1, B1, C1, D1] = ccf_from_tf([1 3 2], [2]);
    modelName = 'Chapter16_Lesson4_Simulink_CCF';
    if bdIsLoaded(modelName)
        close_system(modelName, 0);
    end
    new_system(modelName);
    open_system(modelName);
    add_block('simulink/Sources/Step', [modelName '/Step']);
    add_block('simulink/Continuous/State-Space', [modelName '/CCF_StateSpace']);
    add_block('simulink/Sinks/Scope', [modelName '/Scope']);
    set_param([modelName '/CCF_StateSpace'], ...
        'A', mat2str(A1), 'B', mat2str(B1), 'C', mat2str(C1), 'D', mat2str(D1));
    add_line(modelName, 'Step/1', 'CCF_StateSpace/1');
    add_line(modelName, 'CCF_StateSpace/1', 'Scope/1');
    set_param(modelName, 'StopTime', '10');
    save_system(modelName);
    fprintf('\nSimulink model saved as %s.slx\n', modelName);
end

function [A, B, C, D] = ccf_from_tf(den_desc, num_desc)
    if abs(den_desc(1)) < 1e-14
        error('Leading denominator coefficient must be nonzero.');
    end

    lead = den_desc(1);
    den_desc = den_desc / lead;
    num_desc = num_desc / lead;
    n = length(den_desc) - 1;

    if length(num_desc) > n
        error('Only strictly proper systems are handled.');
    end

    A = zeros(n, n);
    if n > 1
        A(1:n-1, 2:n) = eye(n - 1);
    end
    A(n, :) = -fliplr(den_desc(2:end));

    B = zeros(n, 1);
    B(n) = 1;

    padded = [zeros(1, n - length(num_desc)), num_desc];
    C = fliplr(padded);
    D = 0;
end

function Wc = controllability_matrix(A, B)
    n = size(A, 1);
    Wc = zeros(n, n);
    Ak = eye(n);
    for k = 1:n
        Wc(:, k) = Ak * B;
        Ak = A * Ak;
    end
end

File: Tutorials/Codes/Modern_Control_Course/Chapter16/Lesson4/Chapter16_Lesson4.py
#!/usr/bin/env python3
"""
Chapter16_Lesson4.py

Example systems in controllable canonical form (CCF) for continuous-time SISO
state-space systems.

The convention used here is

    d(s) = s^n + a_{n-1}s^{n-1} + ... + a_1 s + a_0
    n(s) = b_{n-1}s^{n-1} + ... + b_1 s + b_0

with

    A = [[0, 1, 0, ..., 0],
         [0, 0, 1, ..., 0],
         ...
         [-a0, -a1, ..., -a_{n-1}]]

    B = [0, 0, ..., 1]^T
    C = [b0, b1, ..., b_{n-1}]
    D = 0

so that G(s) = C(sI-A)^(-1)B = n(s)/d(s).
"""

from __future__ import annotations

import numpy as np


def controllable_canonical_form(den_desc: list[float], num_desc: list[float]):
    """Return A, B, C, D for a strictly proper SISO transfer function."""
    den = np.asarray(den_desc, dtype=float)
    num = np.asarray(num_desc, dtype=float)

    if den.ndim != 1 or num.ndim != 1:
        raise ValueError("den_desc and num_desc must be one-dimensional lists.")
    if abs(den[0]) < 1e-14:
        raise ValueError("Leading denominator coefficient must be nonzero.")

    num = num / den[0]
    den = den / den[0]
    n = len(den) - 1

    if len(num) > n:
        raise ValueError("This script handles strictly proper transfer functions only.")

    num_padded_desc = np.pad(num, (n - len(num), 0), mode="constant")

    A = np.zeros((n, n))
    if n > 1:
        A[:-1, 1:] = np.eye(n - 1)
    A[-1, :] = -den[1:][::-1]

    B = np.zeros((n, 1))
    B[-1, 0] = 1.0

    C = num_padded_desc[::-1].reshape(1, n)
    D = np.array([[0.0]])
    return A, B, C, D


def controllability_matrix(A: np.ndarray, B: np.ndarray) -> np.ndarray:
    """Compute [B AB ... A^(n-1)B]."""
    n = A.shape[0]
    cols = []
    Ak = np.eye(n)
    for _ in range(n):
        cols.append(Ak @ B)
        Ak = A @ Ak
    return np.hstack(cols)


def eval_transfer(A: np.ndarray, B: np.ndarray, C: np.ndarray, D: np.ndarray, s: complex):
    """Evaluate G(s) = C(sI-A)^(-1)B + D at a scalar complex value s."""
    n = A.shape[0]
    return (C @ np.linalg.solve(s * np.eye(n) - A, B) + D)[0, 0]


def print_system(name: str, den_desc: list[float], num_desc: list[float]):
    A, B, C, D = controllable_canonical_form(den_desc, num_desc)
    Wc = controllability_matrix(A, B)

    print("\n" + "=" * 72)
    print(name)
    print("A =\n", A)
    print("B =\n", B)
    print("C =\n", C)
    print("D =\n", D)
    print("Controllability matrix Wc =\n", Wc)
    print("rank(Wc) =", np.linalg.matrix_rank(Wc))
    print("det(Wc)  =", np.linalg.det(Wc))

    test_points = [0.0, -0.5, 1.0 + 0.2j]
    for z in test_points:
        g_ss = eval_transfer(A, B, C, D, z)
        den_val = np.polyval(den_desc, z)
        num_val = np.polyval(num_desc, z)
        g_tf = num_val / den_val
        print(f"s={z!s:>10}: state-space={g_ss: .8f}, transfer-function={g_tf: .8f}")


if __name__ == "__main__":
    print_system("Example 1: G1(s) = 2/(s^2 + 3s + 2)",
                 den_desc=[1, 3, 2],
                 num_desc=[2])

    print_system("Example 2: G2(s) = (s + 2)/(s^3 + 6s^2 + 11s + 6)",
                 den_desc=[1, 6, 11, 6],
                 num_desc=[1, 2])

    print_system("Example 3: G3(s) = (0.5s^2 + 1.5s + 1)/(s^3 + 4s^2 + 5s + 2)",
                 den_desc=[1, 4, 5, 2],
                 num_desc=[0.5, 1.5, 1.0])
